### PR TITLE
backport: bitcoin#19668, #21598, #19979, #28774 - lock annotations

### PR DESCRIPTION
**Pull request by knst**
## What was done?
These 4 backports improve noticeable implementation of thread-safety analysis by moving many possible warnings to compilation level.
There's a lot of fixes for annotations after that!

## How Has This Been Tested?
Build with clang:
```
CC=clang CXX=clang++ ./configure --prefix=$(pwd)/depends/x86_64-pc-linux-gnu --enable-suppress-external-warnings   --enable-debug  --enable-stacktraces --enable-werror --enable-crash-hooks --enable-maintainer-mode
```

## Breaking Changes
n/a

## Checklist:
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have made corresponding changes to the documentation
- [x] I have assigned this pull request to a milestone

### CODE REVIEW

**Comment by knst** [comment]
@PastaPastaPasta any ideas how to make SharedMutex works again?

See 9c612292d9da768adfa245b63b80007265e24b52

**Comment by knst** [comment]
> PastaPastaPasta any ideas how to make SharedMutex works again?
> See https://github.com/dashpay/dash/commit/9c612292d9da768adfa245b63b80007265e24b52

nevermind, fix a trivial, it is related to new code, not old code.

File: src/util/system.cpp [hunk 1]
```diff
@@ -420,7 +420,7 @@ std::optional<unsigned int> ArgsManager::GetArgFlags(const std::string& name) co
     return std::nullopt;
 }
 
-const fs::path& ArgsManager::GetBlocksDirPath() const
+const fs::path ArgsManager::GetBlocksDirPath() const
```
Comment (UdjinM6) [inline src/util/system.cpp:423]: b80d0ab957655d28bbe495c421c147e3e601fbf6: backport 29040

File: src/net.cpp [hunk 2]
```diff
@@ -643,8 +643,6 @@ CNode* CConnman::ConnectNode(CAddress addrConnect, const char *pszDest, bool fCo
 
 void CNode::CloseSocketDisconnect(CConnman* connman)
 {
-    AssertLockHeld(connman->m_nodes_mutex);
```
Comment (UdjinM6) [inline src/net.cpp:646]: 801545adf56f8fe3f01ee7df48a2e54a0d6206f2: why is it impossible?

File: src/coinjoin/client.cpp [hunk 3]
```diff
@@ -1448,12 +1448,13 @@ bool CCoinJoinClientSession::MakeCollateralAmounts()
...
-        return true;
+    for (const auto& item : vecTally) {
+        if (MakeCollateralAmounts(item, false)) return true;
     }
+
     // There should be at least some denominated funds we should be able to break in pieces to continue mixing
-    if (std::any_of(vecTally.begin(), vecTally.end(), [&](const auto& item) { return MakeCollateralAmounts(item, true); })) {
-        return true;
+    for (const auto& item : vecTally) {
+        if (MakeCollateralAmounts(item, true)) return true;
```
Comment (UdjinM6) [inline src/coinjoin/client.cpp:1457]: 9619f1a708a387aec19864b108d2a7b5e3d897dc: looks unrelated

File: src/util/system.cpp (hunk 1 above)
Comment (knst) [reply src/util/system.cpp:423]: 29040 has too many extra changes that are early to do, should I do as backported?

File: src/net.cpp (hunk 2 above)
Comment (knst) [reply src/net.cpp:646]: Sorry, wrong wording, you don't need this lock at all

File: src/coinjoin/client.cpp (hunk 3 above)
Comment (knst) [reply src/coinjoin/client.cpp:1457]: It looks like unrelated, but actually related because std:: algorithms doesn't pass annotations and it causes compilation errors

File: src/util/system.cpp (hunk 1 above)
Comment (UdjinM6) [reply src/util/system.cpp:423]: I mean, maybe rename it to "partial merge 29040"?

File: src/coinjoin/client.cpp (hunk 3 above)
Comment (UdjinM6) [reply src/coinjoin/client.cpp:1457]: could use ranges here
```
    // First try to use only non-denominated funds
    if (ranges::any_of(vecTally, [&](const auto& item) EXCLUSIVE_LOCKS_REQUIRED(m_wallet.cs_wallet) {
            return MakeCollateralAmounts(item, false);
        })) {
        return true;
    }
    // There should be at least some denominated funds we should be able to break in pieces to continue mixing
    if (ranges::any_of(vecTally, [&](const auto& item) EXCLUSIVE_LOCKS_REQUIRED(m_wallet.cs_wallet) {
            return MakeCollateralAmounts(item, true);
        })) {
        return true;
    }
```

**Review by UdjinM6 — APPROVED** [review]
utACK 1db9e6ac76d895d5e80aba7bf0f0a42d13a904e6

**Review by PastaPastaPasta — APPROVED** [review]
utACK 1db9e6ac76d895d5e80aba7bf0f0a42d13a904e6
